Assignment.Tests/PingProcess.Tests.cs

```diff
@@ -0,0 +1,232 @@
+        cancellationTokenSource.Cancel();
+
+        try
+        {
+            task.Wait();
+            Assert.Fail("Expected an AggregateException to be thrown.");
+        }
+        catch (AggregateException ex)
+        {
+            Assert.IsTrue(ex.InnerExceptions.Any(innerEx => innerEx is TaskCanceledException));
+        }
+    }
+
+    [TestMethod]
+    public async Task RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException()
+    {
+
+        PingProcess pingProcess = new();
+        string hostNameOrAddress = "localhost";
+        using CancellationTokenSource cancellationTokenSource = new();
+
+        Task<PingResult> task = PingProcess.RunAsync(hostNameOrAddress, cancellationTokenSource.Token);
+        cancellationTokenSource.Cancel();
+
+        await Assert.ThrowsExceptionAsync<OperationCanceledException>(() => task);
+    }
+
+    [TestMethod]
+    async public Task RunAsync_MultipleHostAddresses_True()
+    {
+
+        var pingProcess = new PingProcess();
+        string[] hostNames = new string[] { "localhost", "localhost", "localhost", "localhost" };
+
+
+        foreach (var hostName in hostNames)
+        {
+            PingResult result = await PingProcess.RunAsync(hostName);
+
+            Assert.AreEqual(0, result.ExitCode);
+
+            Assert.AreEqual("Success", result.StdOutput);
+        }
+    }
+
+    [TestMethod]
+    public void RunLongRunningAsync_Success()
+    {
+        var startInfo = new ProcessStartInfo
+        {
+            FileName = "ping",
+            Arguments = "localhost -n 5",
+            RedirectStandardOutput = true,
+            RedirectStandardError = true,
+            UseShellExecute = false,
+            CreateNoWindow = true
+        };
+
+        StringBuilder outputBuilder = new();
+        StringBuilder errorBuilder = new();
+
+        Action<string?> outputHandler = (output) => outputBuilder.AppendLine(output);
+        Action<string?> errorHandler = (error) => errorBuilder.AppendLine(error);
+
+        PingProcess ping = new();
+        CancellationTokenSource tokenSource = new();
+
+        var task = PingProcess.RunLongRunningAsync(startInfo, outputHandler, errorHandler, tokenSource.Token);
+
+        var exitCode = task.Result;
+
+        Assert.AreEqual(0, exitCode);
+        Assert.IsFalse(string.IsNullOrEmpty(outputBuilder.ToString()));
+
+
+    }
+
+
+
+    [TestMethod]
+    public void StringBuilderAppendLine_InParallel_IsNotThreadSafe()
+    {
+        IEnumerable<int> numbers = Enumerable.Range(0, short.MaxValue);
+        StringBuilder stringBuilder = new();
+
+        try
+        {
+            numbers.AsParallel().ForAll(item => stringBuilder.AppendLine(""));
+        }catch (AggregateException) { }
+        int lineCount = stringBuilder.ToString().Split(Environment.NewLine).Length;
+        Assert.AreNotEqual(lineCount, numbers.Count() + 1);
+    }
```

Comment:
```suggestion
        try
        {
            numbers.AsParallel().ForAll(item => stringBuilder.AppendLine(""));
                    int lineCount = stringBuilder.ToString().Split(Environment.NewLine).Length;
        Assert.AreNotEqual(lineCount, numbers.Count() + 1);
        }catch (AggregateException) { }

    }
```
This should be here I believe.